### PR TITLE
Update ui functions

### PR DESCRIPTION
**Pull request by malcolmbarrett**
This is the first of two PRs that I plan to submit to address #62. Doesn't quite close that issue.

This PR:
* Updates all user communication to use `ui_*()` functions
* Refactors `ui_*()` functions to pipe through `rlang::inform()` and to use the new internal function `bullet()` to format text via cli.
* Refactors `ui_throw()` to simplify its logic and use `rlang::abort()` to throw an error of class `rextendr_error`
* Renames and modifies `ui_q()` to `ui_o()` to be more similar to `usethis::ui_todo()`, which is usually how questions are asked in usethis
* Allows global UI silencing via `usethis.quiet = FALSE`. I did this to be consistent with usethis, although you may prefer a custom option like `rextendr.quiet`.
* Updates tests. Generally, these are simplified versions of previous message tests since unicode is a nightmare to handle in regex. I think they are good enough tests for their purposes. I also introduced `expect_rextendr_error()` to more strictly expect an error thrown by rextendr.
* I also ran `use_tidy_description()` 🧹 😄 

Shoutout to @Ilia-Kosenkov who laid the groundwork for this PR!

### CODE REVIEW

File: R/register_extendr.R [hunk 1]
```diff
@@ -38,7 +38,7 @@ register_extendr <- function(path = ".", quiet = FALSE, force = FALSE, compile =
       "Unable to register the extendr module.",
       c(
         ui_x("Could not find file {.file src/entrypoint.c }."),
-        ui_q("Are you sure this package is using extendr Rust code?")
+        ui_o("Are you sure this package is using extendr Rust code?")
```
Comment (Ilia-Kosenkov) [inline R/register_extendr.R:41]: Should this be some version of `bullet()` instead of `ui_*`?

**Comment by clauswilke** [comment]
Could you add a file with developer guidelines summarizing how these functions should be used? Along the lines of this file in extendr: https://github.com/extendr/extendr/blob/master/MAINTAINERS_GUIDE.md

File: R/ui.R [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
+}
+
+ui_bullet <- function(type, text, env = parent.frame()) {
+  if (getOption("usethis.quiet", FALSE)) {
+    return(invisible())
+  }
+
+  rlang::inform(bullet(type, text, env = env))
+}
+
```
[the remark this answers is not in the document]
Comment (malcolmbarrett) [reply R/ui.R:22]: I considered that approach, too, so definitely fine with it!

File: R/register_extendr.R (hunk 1 above)
Comment (malcolmbarrett) [reply R/register_extendr.R:41]: yes, missed these -- thanks!

**Comment by malcolmbarrett** [comment]
@clauswilke what do you think about putting it in a `CONTRIBUTING.md` generated by `usethis::use_tidy_contributing()`?

**Comment by clauswilke** [comment]
We have contributing guidelines over at extendr and they could be adapted:
https://github.com/extendr/extendr/blob/master/CONTRIBUTING.md

However, I think that's not the place for documentation that would cover internal coding conventions. I'm not sure where that would go, though. Most projects don't have such documentation and it's often a major problem, as nobody knows how exactly code should be written.

**Comment by malcolmbarrett** [comment]
Yeah, agreed. usethis has a helpful document along those lines here: https://github.com/r-lib/usethis/blob/master/principles.md and in fact has a [section on communicating with the user](https://github.com/r-lib/usethis/blob/master/principles.md#communicating-with-the-user)

**Comment by clauswilke** [comment]
Yes, exactly, something like this.

**Comment by malcolmbarrett** [comment]
Ok, I added a `principles.md` document that includes details on UI and errors. I'll file an issue to include other items outside the scope of this PR

**Review by clauswilke — COMMENTED** [review]
The `principles.md` file looks good to me. I haven't looked over all the other changes as I'm super busy right now.

@Ilia-Kosenkov Can you finish this off?

File: principles.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+This guide documents important internal coding conventions used in rextendr.
+
+## Communicating with the user
+
+rextendr uses the cli package to format messages to the user, which are generally then routed through `rlang::inform()`. Most UI will happen through `ui_*()` functions:
+
+| function | purpose                                                                                                       |
+|----------|---------------------------------------------------------------------------------------------------------------|
+| `ui_v()` | communicate that rextendr has done something, such as write a file                                            |
```
Comment (Ilia-Kosenkov) [inline principles.md:11]: I suggest adding "... has done something *successfully*, ..." to emphasize that `*_v` functions communicate positive results.

Comment (malcolmbarrett) [reply principles.md]: yes, good point -- updated!

**Review by Ilia-Kosenkov — APPROVED** [review]
Looks good to me. Thanks!

**Comment by clauswilke** [comment]
One more thing that I can't check easily from my phone right now: did you add the principles file to .Rbuildignore?

**Comment by malcolmbarrett** [comment]
yes, already done

**Comment by Ilia-Kosenkov** [comment]
Is there anything blocking this PR?

**Comment by clauswilke** [comment]
Not from my end.

**Comment by Ilia-Kosenkov** [comment]
@malcolmbarrett , you should probably merge this.

**Comment by malcolmbarrett** [comment]
I can't. I don't have access 😎

**Comment by Ilia-Kosenkov** [comment]
@malcolmbarrett, oh shoot, I had no idea, I am sorry. If you think it is ready, I can merge it.

**Comment by malcolmbarrett** [comment]
No worries! Yes, I think it is ready

**Comment by malcolmbarrett** [comment]
Thanks!

**Pull request merged**